modules/highgui/src/roiSelector.cpp

```diff
@@ -0,0 +1,193 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+#include <opencv2/imgproc.hpp>
+#include <algorithm>
+
+using namespace cv;
+
+namespace
+{
+class ROISelector
+{
+  public:
+    Rect select(const String &windowName, Mat img, bool showCrossair = true, bool fromCenter = true)
+    {
+        // show notice to user
+        printf("Select a ROI and then press SPACE or ENTER button!\n");
+        printf("Cancel the selection process by pressing c button!\n");
+
+        key = 0;
+        imageSize = img.size();
+
+        // set the drawing mode
+        selectorParams.drawFromCenter = fromCenter;
+
+        // show the image and give feedback to user
+        imshow(windowName, img);
+
+        // copy the data, rectangle should be drawn in the fresh image
+        selectorParams.image = img.clone();
+
+        // select the object
+        setMouseCallback(windowName, mouseHandler, (void *)&selectorParams);
```

Comment:
1) Reliable callback cleanup process is required. Probably with save/restore of previous callback state.
2) Passed `(void *)&selectorParams`, but callback wrapper expects `(void*)this` - add just one "virtual" function or extra field and you would have interesting debug session of crash dumps.
3) `param` should be eliminated in this non-static method:
```
    void opencv_mouse_callback(int event, int x, int y, int, void *param)
    {
        handlerT *data = (handlerT *)param;
```
4. 'ESC' key inconsistency. Some highgui backends may return capital 'C' symbol.
5. 1000 FPS processing is not required (`waitKey(1)`). 20 fps is enough.
6. `& 0xff` is not required in OpenCV 3.2+ .